Upload checksums for installation artifacts

naamunds opened the pull request:
The checksums are SHA-512 hashes, which users can use to verify file
integrity and authenticity.

Fixes #3693.

cc: @livarcocc @blackdwarf @MichaelSimons

naamunds commented:
This PR should not be merged until environment variables have been added to the build definitions for `CHECKSUM_STORAGE_KEY`, `CHECKSUM_STORAGE_ACCOUNT`, and `CHECKSUM_STORAGE_CONTAINER`.  This is blocked by setting up the new Azure Storage account for the checksums, which is still in progress.

MichaelSimons commented:
>   <UsingTask TaskName="UploadToAzure" AssemblyFile="$(CLIBuildDll)"/>

Does this need to be updated/broken into two tasks, UploadChecksumsToAzure and UploadArtifactsToAzure?

---
Refers to: build/Microsoft.DotNet.Cli.tasks:28 in 22fd92a. [](commit_id = 22fd92a69017a6eb0ede189d368f77b4b99f86e1, deletion_comment = False)

MichaelSimons commented on scripts/dockerrun.sh:
```diff
@@ -129,5 +129,8 @@ docker run $INTERACTIVE -t --rm --sig-proxy=true \
     -e STORAGE_KEY \
     -e STORAGE_ACCOUNT \
     -e STORAGE_CONTAINER \
```
These names are somewhat ambiguous given the new checksum variables.  Would it make sense to qualify them as for what they are intended to be used for (e.g. artifacts)?

brthor replied:
👍 

You'll need to change the names in the build definitions as well

naamunds replied:
Fixed.  I'll update the build definition too right before merging this PR.

MichaelSimons commented:
:shipit:

livarcocc commented on build_projects/dotnet-cli-build/GenerateChecksums.cs:
```diff
@@ -0,0 +1,61 @@
+                        throw new Exception($"The file '{item.ItemSpec}' does not exist.");
+                    }
+
+                    Log.LogMessage(MessageImportance.High, "Generating checksum for '{0}' into '{1}'...", item.ItemSpec, destinationPath);
```
NIT: can you try to keep the length of this line within 120 columns? No horizontal scrolling is nice.

brthor replied:
It is hard to do in msbuild at times but 👍 
Long lines suck to review

naamunds replied:
Agreed. Fixed.

livarcocc commented:
@dotnet-bot Test OSX x64 Release Build
@dotnet-bot Test RHEL7.2 x64 Release Build

brthor commented on build/publish/Microsoft.DotNet.Cli.Checksum.targets:
```diff
@@ -0,0 +1,21 @@
+                            SetupGenerateChecksumsInputs;"
+          Inputs="@(GenerateChecksumsInputs)"
+          Outputs="%(GenerateChecksumsInputs.DestinationPath)">
+    <GenerateChecksums Items="@(GenerateChecksumsInputs)" />
```
nit: 
`GenerateChecksumsInputs` doesn't tell me much about the contents of these inputs. 

Can you rename this property to reflect what type of inputs it is?

naamunds replied:
I've renamed it to `ArtifactsForGeneratingChecksums`, and renamed the target which initializes it to `InitializeChecksumItemGroup`.

brthor replied:
Nice 👍

brthor commented on build/publish/PublishContent.targets:
```diff
@@ -38,4 +38,23 @@
+      AccountKey="$(ChecksumCloudDropAccessToken)"
+      AccountName="$(ChecksumCloudDropAccountName)"
+      ContainerName="$(ChecksumContainerName)"
+      Items="@(GenerateChecksumsInputs->'%(DestinationPath)')"
```
Shouldn't this be consuming the GenerateChecksums Outputs rather than the inputs?

naamunds replied:
The `DestinationPath` metadata represents the outputs.  I think this is less confusing now that the item type has been renamed to `ArtifactsForGeneratingChecksums`:
`Items="@(ArtifactsForGeneratingChecksums->'%(DestinationPath)')"`

brthor replied:
Agreed it is less confusing now

brthor commented on build_projects/dotnet-cli-build/GenerateChecksums.cs:
```diff
@@ -0,0 +1,61 @@
+        /// 'DestinationPath' that specifies the path of the checksum file to create.
+        /// </summary>
+        [Required]
+        public ITaskItem[] Items { get; set; }
```
Rather than relying on this `DestinationPath` metadata to be present can we use the same semantics as the copy task, where there are two input lists, source files and destination files, with a 1:1 mapping between the two

naamunds replied:
I thought about implementing it like that, but decided to follow the pattern used by the `UploadToAzure` and `SetBlobPropertiesBasedOnFileType` tasks, which each rely on `RelativeBlobPath` metadata being set.  I think the documentation on the Items variable plus the error message if it's not set make it discoverable enough for clients.

brthor replied:
Fair enough.

brthor commented on build_projects/dotnet-cli-build/GenerateChecksums.cs:
```diff
@@ -0,0 +1,61 @@
+                    {
+                        HashAlgorithm hashAlgorithm = SHA512.Create();
+                        byte[] hash = hashAlgorithm.ComputeHash(stream);
+                        string checksum = BitConverter.ToString(hash).Replace("-", string.Empty);
```
It feels strange that the hash is being mutated here, why are you removing the '-'?

Does the hash still match the output of other sha512 generators?

brthor replied:
I see this is because of the bit converter outputting hex pairs.

Rather than returning the raw hex data then would it make sense to use something like base64 encoding?
https://msdn.microsoft.com/en-us/library/system.convert.tobase64string(v=vs.110).aspx

naamunds replied:
From what I've seen, hash values are typically written as hexadecimal strings without hyphens.  This is the format used by PowerShell's `Get-FileHash` cmdlet, in the Wikipedia article on SHA-2, and in the checksums provided by WinSCP (which I just happened to see today when downloading it).

brthor replied:
I just checked the output of openssl on mac as well and it does look to be the hex string. 

Sounds good

naamunds commented:
I've added a commit to upload "latest" copies of the checksums, as is already done with the installation artifacts.  In doing so, I refactored the FinishBuild logic to have more atomic build tasks (`CheckIfAllBuildsHavePublished`, `CopyBlobsToLatest`, and `UpdateVersionsRepo`), because it seemed like with uploading the "latest" checksums, the `FinalizeBuild` task would have too many responsibilities.

naamunds commented:
The CI builds for this PR were failing because the calls to the `AzurePublisher` constructor in `CheckIfAllBuildsHavePublished` and `CopyBlobsToLatest` depend on the changes in #4234.  (I had meant to make the two PRs independent but missed a couple spots.)  I've added a commit to remove this.  This means that the container names will be hard-coded to "dotnet" until #4234 is merged.

naamunds commented:
I just received the new Storage Account info, so I'm ready to merge this PR after updating the VSTS build definition environment variables and after the Jenkins checks complete.

Now that #4234 has been merged, I've added back the ContainerName changes here which I had had to remove.

I'll squash these changes into one commit before merging.

naamunds commented:
@dotnet-bot Test Windows_NT x64 Release Build
